File: src/main/java/hacker/solution/CovariantDemo.java
package hacker.solution;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

//Complete the classes below
class Flower {
    String whatsYourName() {
        return "I have many names and types.";
    }
}

class Sunflower extends Flower {
    @Override
    String whatsYourName() {
        return "Sunflower";
    }
}

class Jasmine extends Flower {
    @Override
    String whatsYourName() {
        return "Jasmine";
    }
}

class Lily extends Flower {
    @Override
    String whatsYourName() {
        return "Lily";
    }
}

class Region {
    Flower yourNationalFlower() {
        return new Flower();
    }
}

class Karnataka extends Region {
    @Override
    Flower yourNationalFlower() {
        return new Sunflower();
    }
}

class WestBengal extends Region {
    @Override
    Flower yourNationalFlower() {
        return new Jasmine();
    }
}

class AndhraPradesh extends Region {
    @Override
    Flower yourNationalFlower() {
        return new Lily();
    }
}

public class CovariantDemo {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        String s = reader.readLine().trim();
        Region region = null;
        switch (s) {
            case "WestBengal":
                region = new WestBengal();
                break;
            case "AndhraPradesh":
                region = new AndhraPradesh();
                break;
            case "Karnataka":
                region = new Karnataka();
                break;
        }
        Flower flower = region.yourNationalFlower();
        System.out.println(flower.whatsYourName());
    }
}

File: src/main/kotlin/hacker/solutionk/CovariantDemo.kt
package hacker.solutionk

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

open class Flower {
    open fun whatsYourName(): String {
        return "I have many names and types."
    }
}

internal class Sunflower : Flower() {
    override fun whatsYourName(): String {
        return "Sunflower"
    }
}

internal class Jasmine : Flower() {
    override fun whatsYourName(): String {
        return "Jasmine"
    }
}

internal class Lily : Flower() {
    override fun whatsYourName(): String {
        return "Lily"
    }
}

open class Region {
    open fun yourNationalFlower(): Flower {
        return Flower()
    }
}

internal class Karnataka : Region() {
    override fun yourNationalFlower(): Flower {
        return Sunflower()
    }
}

internal class WestBengal : Region() {
    override fun yourNationalFlower(): Flower {
        return Jasmine()
    }
}

internal class AndhraPradesh : Region() {
    override fun yourNationalFlower(): Flower {
        return Lily()
    }
}

object CovariantDemo {
    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val reader = BufferedReader(InputStreamReader(System.`in`))
        val s = reader.readLine().trim { it <= ' ' }
        var region: Region? = null
        when (s) {
            "WestBengal" -> region = WestBengal()
            "AndhraPradesh" -> region = AndhraPradesh()
            "Karnataka" -> region = Karnataka()
        }
        val flower = region!!.yourNationalFlower()
        println(flower.whatsYourName())
    }
}